src/tools/rbd_mirror/group_replayer/Replayer.cc

```diff
@@ -1146,63 +1206,101 @@ bool Replayer<I>::prune_all_image_snapshots(cls::rbd::GroupSnapshot *local_snap)
     }
-    std::unique_lock locker{m_lock};
     auto snap_type = cls::rbd::get_group_snap_namespace_type(
         local_snap->snapshot_namespace);
-    if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_USER) {
-      bool unlink_user_snap = true;
-      for (auto &remote_snap : m_remote_group_snaps) {
-        if (remote_snap.name == local_snap->name) {
-          unlink_user_snap = false;
-          break;
-        }
-      }
-      if (!unlink_user_snap) {
-        locker.unlock();
-        continue;
-      }
-      dout(10) << "unlinking regular group snap in-progress: "
-               << local_snap->name << ", with id: " << local_snap->id << dendl;
-    } else {
+    if (snap_type == cls::rbd::GROUP_SNAPSHOT_NAMESPACE_TYPE_MIRROR) {
       // remove mirror peer uuid of the remote snap
-
       auto next_local_snap = std::next(local_snap);
       // If next local snap is end, or if it is the syncing in-progress snap,
       // then we still need this group snapshot.
       if (next_local_snap == m_local_group_snaps.end() ||
-          (next_local_snap->id == last_local_snap_id &&
+          (next_local_snap->id == last_local_mirror_snap_id &&
            next_local_snap->state != cls::rbd::GROUP_SNAPSHOT_STATE_COMPLETE)) {
         break;
       }
       remove_mirror_peer_uuid(local_snap->id);
     }
+  }
+
+  if (!m_prune_snap_ids.empty()) {
+    for (auto prune_snap_id: m_prune_snap_ids) {
+      prune_group_snapshot(prune_snap_id, locker);
+    }
+  }
+}
+
+template <typename I>
+void Replayer<I>::prune_group_snapshot(std::string snap_id,
+    std::unique_lock<ceph::mutex> &locker) {
+  ceph_assert(ceph_mutex_is_locked_by_me(m_lock));
+  dout(10) << "group snapshot id: " << snap_id << dendl;
+
+  for (auto local_snap = m_local_group_snaps.begin();
+      local_snap != m_local_group_snaps.end(); ++local_snap) {
+    auto local_snap_id = local_snap->id;
+    if (local_snap_id != snap_id) {
+      continue;
+    }
     locker.unlock();
+
     // prune all the image snaps of the group snap locally
     if (prune_all_image_snapshots(&(*local_snap))) {
+      locker.lock();
       continue;
     }
-    dout(10) << "all image snaps are pruned, finally unlinking group snap: "
-             << local_snap->id << dendl;
+    dout(10) << "all image snaps are pruned, finally prune group snapshot: "
+             << local_snap_id << dendl;
     locker.lock();
-    r = librbd::cls_client::group_snap_remove(&m_local_io_ctx,
-        librbd::util::group_header_name(m_local_group_id), local_snap->id);
+    int r = librbd::cls_client::group_snap_remove(&m_local_io_ctx,
+        librbd::util::group_header_name(m_local_group_id), local_snap_id);
     if (r < 0) {
       derr << "failed to remove group snapshot : "
-           << local_snap->id << " : " << cpp_strerror(r) << dendl;
+           << local_snap_id << " : " << cpp_strerror(r) << dendl;
     }
     m_local_group_snaps.erase(local_snap);
-    locker.unlock();
+    auto iter = std::find_if(
```

Comment:
m_prune_snap_ids is being accessed without a lock. It can also be modified by the call to `handle_remove_mirror_peer_uuid`.